packages/go_router/lib/src/route.dart

```diff
@@ -433,22 +462,339 @@ class ShellRoute extends RouteBase {
+///   routes: <RouteBase>[
+///     StatefulShellRoute(
+///       builder: (BuildContext context, GoRouterState state,
+///           Widget navigationContainer) {
+///         return ScaffoldWithNavBar(body: navigationContainer);
+///       },
+///       pageBuilder:
+///           (BuildContext context, GoRouterState state, Widget statefulShell) {
+///         return NoTransitionPage<dynamic>(child: statefulShell);
+///       },
+///       branches: <StatefulShellBranch>[
+///         /// The first branch, i.e. root of tab 'A'
+///         StatefulShellBranch(routes: <RouteBase>[
+///           GoRoute(
+///             parentNavigatorKey: _tabANavigatorKey,
+///             path: '/a',
+///             builder: (BuildContext context, GoRouterState state) =>
+///               const RootScreen(label: 'A', detailsPath: '/a/details'),
+///           ),
+///         ]),
+///         /// The second branch, i.e. root of tab 'B'
+///         StatefulShellBranch(routes: <RouteBase>[
+///           GoRoute(
+///             parentNavigatorKey: _tabBNavigatorKey,
+///             path: '/b',
+///             builder: (BuildContext context, GoRouterState state) =>
+///               const RootScreen(label: 'B', detailsPath: '/b/details/1'),
+///           ),
+///         ]),
+///       ],
+///     ),
+///   ],
+/// );
+/// ```
+///
+/// To access the current state of this route, to for instance access the
+/// index of the current route branch - use the method
+/// [StatefulShellRouteState.of]. For example:
+///
+/// ```
+/// final StatefulShellRouteState shellState = StatefulShellRouteState.of(context);
+/// ```
+///
+/// See [Stateful Nested Navigation](https://github.com/flutter/packages/blob/main/packages/go_router/example/lib/stateful_nested_navigation.dart)
+/// for a complete runnable example using StatefulShellRoute.
+class StatefulShellRoute extends ShellRouteBase {
+  /// Constructs a [StatefulShellRoute] from a list of [StatefulShellBranch]es,
+  /// each representing a separate nested navigation tree (branch).
   ///
-  /// Similar to GoRoute builder, but with an additional child parameter. This
-  /// child parameter is the Widget built by calling the matching sub-route's
-  /// builder.
-  final ShellRouteBuilder? builder;
+  /// A separate [Navigator] will be created for each of the branches, using
+  /// the navigator key specified in [StatefulShellBranch]. Note that unlike
+  /// [ShellRoute], a builder must always be provided when creating a
+  /// StatefulShellRoute. The pageBuilder however is optional, and is used
+  /// in addition to the builder.
+  StatefulShellRoute({
+    required this.branches,
+    required super.builder,
+    super.pageBuilder,
+  })  : assert(branches.isNotEmpty),
+        assert(_debugUniqueNavigatorKeys(branches).length == branches.length,
+            'Navigator keys must be unique'),
+        super._(routes: _routes(branches)) {
+    for (int i = 0; i < routes.length; ++i) {
```

Comment:
Can this entire for loop be a static method in the initializer? and it should be wrapped in assert as well

Reply:
It can - will fix.